File: app/src/main/java/io/inabsentia/superhangman/data/dao/MatchDAO.java
package io.inabsentia.superhangman.data.dao;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import io.inabsentia.superhangman.data.dto.MatchDTO;

import static android.content.ContentValues.TAG;

public class MatchDAO implements IMatchDAO {

    private static MatchDAO instance = new MatchDAO();

    private List<MatchDTO> matchList;

    private MatchDAO() {
        matchList = new ArrayList<>();
    }

    static {
        try {
            instance = new MatchDAO();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static synchronized MatchDAO getInstance() {
        return instance;
    }

    @Override
    public void add(MatchDTO dto) throws DALException {
        matchList.add(dto);
    }

    @Override
    public MatchDTO get(int id) throws DALException {
        for (MatchDTO dto : matchList)
            if (dto.getId() == id) return dto;

        throw new DALException("Id [" + id + "] not found!");
    }

    @Override
    public List<MatchDTO> getAll() throws DALException {
        if (matchList == null) throw new DALException("MatchList is null!");
        return matchList;
    }

    @Override
    public void update(MatchDTO dto) throws DALException {
        MatchDTO existingDTO = get(dto.getId());
        matchList.remove(existingDTO);
        matchList.add(dto);
    }

    @Override
    public void remove(int id) throws DALException {
        MatchDTO highScoreDTO = get(id);
        matchList.remove(highScoreDTO);
    }

    @Override
    public void removeAll() throws DALException {
        matchList = new ArrayList<>();
    }

    @Override
    public void load(Context context) throws DALException {
        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);
        Gson gson = new Gson();
        String json = sharedPrefs.getString(TAG, null);
        Type type = new TypeToken<ArrayList<MatchDTO>>() {
        }.getType();
        matchList = gson.fromJson(json, type);

        if (matchList == null) {
            matchList = new ArrayList<>();
            save(context);
        }
    }

    @Override
    public void save(Context context) throws DALException {
        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);
        SharedPreferences.Editor editor = sharedPrefs.edit();
        Gson gson = new Gson();
        String json = gson.toJson(matchList);
        editor.putString(TAG, json);
        editor.apply();
    }

}

File: app/src/main/java/io/inabsentia/superhangman/data/dao/MatchDAO.kt
package io.inabsentia.superhangman.data.dao

import android.content.ContentValues.TAG
import android.content.Context
import android.preference.PreferenceManager
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import io.inabsentia.superhangman.data.dto.MatchDTO
import java.util.*

class MatchDAO private constructor() : IMatchDAO {

    private var matchList: MutableList<MatchDTO>? = null

    override val all: List<MatchDTO>
        @Throws(DALException::class)
        get() {
            if (matchList == null) throw DALException("MatchList is null!")
            return matchList as MutableList<MatchDTO>
        }

    init {
        matchList = ArrayList()
    }

    @Throws(DALException::class)
    override fun add(dto: MatchDTO) {
        matchList!!.add(dto)
    }

    @Throws(DALException::class)
    override fun get(id: Int): MatchDTO {
        matchList!!.filter { it.id == id }.forEach { return it }
        throw DALException("Id [$id] not found!")
    }

    @Throws(DALException::class)
    override fun update(dto: MatchDTO) {
        val existingDTO = get(dto.id)
        matchList!!.remove(existingDTO)
        matchList!!.add(dto)
    }

    @Throws(DALException::class)
    override fun remove(id: Int) {
        val highScoreDTO = get(id)
        matchList!!.remove(highScoreDTO)
    }

    @Throws(DALException::class)
    override fun removeAll() {
        matchList = ArrayList()
    }

    @Throws(DALException::class)
    override fun load(context: Context) {
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)
        val gson = Gson()
        val json = sharedPrefs.getString(TAG, null)
        val type = object : TypeToken<ArrayList<MatchDTO>>() {}.type

        matchList = gson.fromJson<List<MatchDTO>>(json, type) as MutableList<MatchDTO>?

        if (matchList == null) {
            matchList = ArrayList()
            save(context)
        }
    }

    @Throws(DALException::class)
    override fun save(context: Context) {
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)
        val editor = sharedPrefs.edit()
        val gson = Gson()
        val json = gson.toJson(matchList)
        editor.putString(TAG, json)
        editor.apply()
    }

    companion object {
        @get:Synchronized
        var instance: IMatchDAO? = null
            private set

        init {
            instance = MatchDAO()
        }
    }

}